### ISSUE TITLE
Application settings are not overriding tenant settings for JWT token lifetimes

### ISSUE THREAD

**Issue by gurupras** [FusionAuth/fusionauth-issues#1032]
### Description 
Application settings for JWT token lifetime is not overriding the tenant setting. This makes it very difficult for me to debug my application code that deals with token expiration and renewal. [See this forum post for more information](https://fusionauth.io/community/forum/topic/604/unexpected-validity-of-access_token-obtained-via-exchanging-refresh_token).

### Affects versions
FusionAuth v1.18.8

### Steps to reproduce
1. Create application with extremely short JWT lifetime
2. Login, get `access_token` and `refresh_token` (use specific `clientId`)
3. Exchange `refresh_token` for a new `access_token` (again, use specific `clientId`)
4. New `access_token` has lifetime specified by tenant rather than application

### Expected behavior
Tokens should have lifetime as defined (if defined) by the application.

### Screenshots
![image](https://user-images.githubusercontent.com/2933207/101786787-56dd1180-3acc-11eb-9011-86ef4f14cf70.png)
![image](https://user-images.githubusercontent.com/2933207/101786795-59d80200-3acc-11eb-86bf-e58c846c38ab.png)

```
Token expires in 28475 {aud: "faa64edb-ecca-4acb-a35e-d83f395ac04e", iat: 1607337214, exp: 1607337244}
Refreshing tokens to avoid expiry {aud: "faa64edb-ecca-4acb-a35e-d83f395ac04e", iat: 1607337214, exp: 1607337244}
Token expires in 35999195 {aud: "faa64edb-ecca-4acb-a35e-d83f395ac04e", iat: 1607337215, exp: 1607373215}
```

### Platform
 - Device: Desktop
 - OS: Ubuntu 20.04
 - Browser + version Google Chrome 87.0.4280.88 (Official Build) (64-bit)
 - Database PostgresSQL v12.5 (I think?)

**Comment by mooreds** [comment]
Is this replicable on 1.22.1?

**Comment by gurupras** [comment]
If the question was for me, I don't know if I can procure that information for you. I don't know how to get that information without disabling auth for our production servers.

**Comment by mooreds** [comment]
@gurupras I am unable to replicate this on 1.18.4 or 1.22.1. I am using the login API instead of the oauth API, but that should follow the same path.

I do the following from a vanilla install of FusionAuth

* create an api key
* create a tenant
* create an application in this tenant
* set the JWT to expire in 5 seconds for the application (tenant has default value of 3600 for jwt duration)
* enable `Generate Refresh Tokens` and `Enable JWT refresh` on the security tab of the application
* Add a user to this new tenant
* Register the user to the application

Call this login.sh shell script:

```
API_KEY=SNjNZj8jz4A_5BeL07pF901nwlLxRQ3CK6shpuIFQkg
APP_ID=55e3590f-bb5c-43a2-a9f1-f729db0aab53
TENANT_ID=32306536-3036-6431-3865-646430303332	

curl -v -XPOST -H 'Content-type: application/json' -H "Authorization: $API_KEY" 'http://localhost:9011/api/login' -H "X-FusionAuth-TenantId: $TENANT_ID" -d '{"loginId": "test@example.com", "password": "password", "applicationId" : "'$APP_ID'" }'
```

Then I get a refresh token back, and I call this script like so: `sh refresh.sh [refreshToken]`. Here's the script:

```
API_KEY=SNjNZj8jz4A_5BeL07pF901nwlLxRQ3CK6shpuIFQkg
APP_ID=55e3590f-bb5c-43a2-a9f1-f729db0aab53
TENANT_ID=32306536-3036-6431-3865-646430303332	

curl -v -XPOST -H 'Content-type: application/json' -H "Authorization: $API_KEY" 'http://localhost:9011/api/jwt/refresh' -H "X-FusionAuth-TenantId: $TENANT_ID" -d '{"refreshToken": "'$1'", "applicationId" : "'$APP_ID'" }'
```

And I get back a JWT with the correct duration (5 s) every time.

Can you try this and see if you get the correct value (which would indicate that my understanding of this being the same path is wrong and I'd need to try to use the oauth path).

**Comment by redondosole** [comment]
Has this been solved?
I am having similar issue, the setting that i am trying to override is the siging key and it does not change.
Easy way to replicate: 
- Default key is HS256. Create an RS256 and assign it to the application. 
Expected:
- The body of the token should be bigger with the new encryption type
- The alg should read RS256.
Actual:
- alg of the token is still HS256

The only way that I get to change the token alg is to change the Tenant JWT settings. We wanted to use a different key for this app tho.

Thanks for any insight anyone might have.

**Comment by jobannon** [comment]
@redondosole I followed your steps and was not able to reproduce.  I was able to select an application-specific 2K RSA key, and then login to that application, and have the KID be used to sign my access token (as I specified in the application level configuration).  

Can you confirm the version of FusionAuth you are running?  Can you confirm what you are trying to sign with this key or any other details?

**Comment by redondosole** [comment]
Hi. thanks for checking this out.

More details:
- There are two tenants, the second tenant has one app assigned. 
The app has 
JWT ->
- Enabled
- Token Signing (picked created with RS256)
- ID signing (picked same RS256
Security->
- Require a key (off)
- Generate Refresh token (on)
- Enable JWT Refresh (on)
Also, I just noticed that the refresh tokens are not coming back.
Is there some sort of cache that needs to be busted?

**Comment by redondosole** [comment]
So we figured out the issue. Other folks might be having the same problem. 

Our issue was that we mixed authentication and authorization. 

When the user logged in with right credentials, belonging to a group but the registration was missing, we were getting a 202 with a payload. The token and its settings were coming from the Tenant, not the application.

Only when the user had the right registration to the app, then the application settings were kicking in, including the different key (HS vs RS). 

Even when the group provides a role to the apps but if the application is not added to the user in the registration, the authorization request would soft-fail with a 202, and provide a tenant token.

We now need to add all user to all apps, with no role, then based on the group they get the access that they need.

Hope this helps someone else with the same issue.

Thanks team for looking into this.
